File: sdk-libs/gnark-sdk/utxo.go
package gnarksdk

import (
	"github.com/consensys/gnark/frontend"

	spp "zolana/prover/circuits/spp_transaction/shared"
)

// Utxo is one UTXO's witness, in the field order and with the field names
// zolana-gnark-ffi-prover's utxo_witness_entries writes. TreeID is the raw u16
// id of the tree the UTXO lives in: the tree an input is spent from, the tree
// an output is appended to.
type Utxo struct {
	Domain        frontend.Variable
	Owner         frontend.Variable
	Asset         frontend.Variable
	Amount        frontend.Variable
	Blinding      frontend.Variable
	DataHash      frontend.Variable
	RingDataHash  frontend.Variable
	RingProgramID frontend.Variable
	TreeID        frontend.Variable
}

// Hash is the UTXO hash the shielded pool commits.
func (u Utxo) Hash(api frontend.API) frontend.Variable {
	return spp.UtxoHashCircuit(api, spp.UtxoCircuitFields{
		Domain:        u.Domain,
		Owner:         u.Owner,
		Asset:         u.Asset,
		Amount:        u.Amount,
		Blinding:      u.Blinding,
		DataHash:      u.DataHash,
		RingDataHash:  u.RingDataHash,
		RingProgramID: u.RingProgramID,
	}, u.TreeID)
}

// AssertDefaultRing asserts u is a spendable UTXO outside every ring. Amount
// and TreeID are not range-checked: the transact proof checks both for every
// UTXO hash it folds into the same private transaction hash.
func (u Utxo) AssertDefaultRing(api frontend.API) {
	api.AssertIsEqual(u.Domain, spp.UtxoDomain)
	api.AssertIsEqual(u.RingDataHash, 0)
	api.AssertIsEqual(u.RingProgramID, 0)
}

File: sdk-libs/gnark-ffi-prover/src/utxo.rs
use zolana_client::ProofInputUtxo;

use crate::decimal;

/// Encodes one UTXO as proof input entries for a Go circuit.
///
/// The nine `{prefix}_{field}` keys are the reflected field names of the
/// circuit's `gnarksdk.Utxo` field, in its declaration order.
pub fn utxo_proof_inputs(utxo: &ProofInputUtxo, prefix: &str) -> Vec<(String, Vec<String>)> {
    let fields: [(&str, &[u8; 32]); 9] = [
        ("Domain", &utxo.domain),
        ("Owner", &utxo.owner_hash),
        ("Asset", &utxo.asset),
        ("Amount", &utxo.amount),
        ("Blinding", &utxo.blinding),
        ("DataHash", &utxo.data_hash),
        ("RingDataHash", &utxo.ring_data_hash),
        ("RingProgramID", &utxo.ring_program_id),
        ("TreeID", &utxo.tree_id),
    ];
    fields
        .iter()
        .map(|(suffix, value)| (format!("{prefix}_{suffix}"), vec![decimal(value)]))
        .collect()
}

/// The key set an encoder must produce for one UTXO prefix.
pub fn utxo_proof_input_keys(prefix: &str) -> Vec<String> {
    [
        "Domain",
        "Owner",
        "Asset",
        "Amount",
        "Blinding",
        "DataHash",
        "RingDataHash",
        "RingProgramID",
        "TreeID",
    ]
    .iter()
    .map(|suffix| format!("{prefix}_{suffix}"))
    .collect()
}
